split: use std::io::copy() with new writer implementation to improve maintainability and speed

jfinkels opened the pull request:
Previously, the `ByteSplitter` and `LineSplitter` structs were responsible for both reading the input and writing it to the current output file. The responsibility of creating a new output file for each chunk lived outside of these structs. By removing the responsibility of reading and only focusing on writing, we can simplify the implementation, minimize the responsibilities of these structs, and see a speed improvement.

This pull request replaces the `ByteSplitter` and `LineSplitter` structs with `ByteChunkWriter` and `LineChunkWriter` structs, respectively. The latter implement `Write`, which allows them to be used with `std::io::copy()`. Conceptually, bytes from the input file are now repeatedly read and sent to the writer, and the writer is responsible for deciding when a new output file needs to be created (and creating that file).

This change has no significant change in the speed of the program in the "split by bytes" mode of operation. I tested this by first creating a file with ten million null bytes:
```
$ printf "%0.s\0" {1..10000000} > testfile.txt
```
and then testing the GNU split, the `master` branch version (at commit cb051e7416f2406a2ebc1d7a58f6e263f5cfcaf2) of `uu_split`, and the new version of `uu_split` in this branch:
```
$ hyperfine --warmup 10 --prepare 'rm x* || true' "split -b 1000 testfile.txt" "./uusplit-master -b 1000 testfile.txt" "./uusplit-new -b 1000 testfile.txt"
Benchmark 1: split -b 1000 testfile.txt
  Time (mean ± σ):     266.1 ms ±  14.6 ms    [User: 4.9 ms, System: 257.2 ms]
  Range (min … max):   253.4 ms … 302.3 ms    10 runs
 
Benchmark 2: ./uusplit-master -b 1000 testfile.txt
  Time (mean ± σ):     117.8 ms ±   3.0 ms    [User: 9.9 ms, System: 106.8 ms]
  Range (min … max):   112.6 ms … 123.8 ms    15 runs
 
Benchmark 3: ./uusplit-new -b 1000 testfile.txt
  Time (mean ± σ):     116.3 ms ±   3.1 ms    [User: 11.1 ms, System: 104.2 ms]
  Range (min … max):   111.0 ms … 121.1 ms    14 runs
 
Summary
  './uusplit-new -b 1000 testfile.txt' ran
    1.01 ± 0.04 times faster than './uusplit-master -b 1000 testfile.txt'
    2.29 ± 0.14 times faster than 'split -b 1000 testfile.txt'
```

This change does however produce about a 1.4 times improvement in speed in the "split by lines" mode of operation. I tested this by getting the first ten million lines of a Wikidata database dump XML file:
```
$ wget -O - https://dumps.wikimedia.org/wikidatawiki/20211001/wikidatawiki-20211001-pages-logging.xml.gz | gunzip | head -n 10000000 > wikidata_10000000.xml
```
and then testing as before (the default mode of operation is "split by lines in chunks of one thousand lines each"):
```
$ hyperfine --warmup 10 --prepare 'rm x* || true' "split wikidata_10000000.xml" "./uusplit-master wikidata_10000000.xml" "./uusplit-new wikidata_10000000.xml"
Benchmark 1: split wikidata_10000000.xml
  Time (mean ± σ):      1.523 s ±  0.628 s    [User: 0.095 s, System: 0.529 s]
  Range (min … max):    0.628 s …  2.414 s    10 runs
 
Benchmark 2: ./uusplit-master wikidata_10000000.xml
  Time (mean ± σ):     654.4 ms ±   7.9 ms    [User: 308.5 ms, System: 345.1 ms]
  Range (min … max):   645.2 ms … 673.0 ms    10 runs
 
Benchmark 3: ./uusplit-new wikidata_10000000.xml
  Time (mean ± σ):     463.1 ms ±  16.8 ms    [User: 114.3 ms, System: 344.1 ms]
  Range (min … max):   450.0 ms … 507.3 ms    10 runs
 
Summary
  './uusplit-new wikidata_10000000.xml' ran
    1.41 ± 0.05 times faster than './uusplit-master wikidata_10000000.xml'
    3.29 ± 1.36 times faster than 'split wikidata_10000000.xml'
```
I have added a brief `BENCHMARKING.md` file describing the general procedure I followed for benchmarking this program.

jfinkels commented:
Can someone advise how to ignore the spellchecker on the `BENCHMARKING.md` file?

tertsdiepraam commented:
Impressive results!

> Can someone advise how to ignore the spellchecker on the BENCHMARKING.md file?

The `wc` benchmarking file does it with an HTML comment:
```html
<!-- spell-checker:ignore (words) uuwc uucat largefile somefile Mshortlines moby lwcm cmds tablefmt -->
```

rivy commented:
> Can someone advise how to ignore the spellchecker on the `BENCHMARKING.md` file?

We're using [`cspell`](https://cspell.org; see [in-document configuration](https://cspell.org/configuration/document-settings)).

Don't turn it off for the document. Instead, use something like `<!-- spell-checker:ignore (words) testfile -->`, preferably near the top of the file, or rename `testfile` to something acceptable to the spell checker (eg, using snake [`test_file`] or camel [`testFile`] casing).

`testfile` with the noted spell-checker exception is fine IMO.

rivy reviewed (changes requested):
Nice results, and I think the coding style looks great!

But there's a new failure for `test/split/line` in the GNU test suite from the code changes. When looking at the raw log, it looks like there are some missing final newlines.

jfinkels commented:
Thanks, I'll take a look at the GNU test failures and add some corresponding unit tests to `test_split.rs`.

jfinkels commented:
I've added a fixup commit that should resolve the line ending issue and I added a unit test with a more detailed test for the contents of the output files to prevent any future regression. (It was an off-by-one issue: I was only writing the bytes up to the newline character but not including the newline character.)

sylvestre commented:
Sorry, we have a conflict, could you please have a look?
thanks

jfinkels commented:
Okay, I have resolved the conflicts and rebased the branch on `master`.

sylvestre commented:
needs a merge too :)

sylvestre commented:
sorry :(

jfinkels commented:
No problem, I produce a lot of self-contained pull requests so I expect to resolve merge conflicts sometimes. I'll focus on this next.
